### ISSUE TITLE
persist: end-to-end read performance benchmark

### ISSUE THREAD

**Issue by danhhz** [MaterializeInc/materialize#7878]
The primary thing we hear again and again from customers that they want from persistence is faster restarts. Before operator persistence (which is a long way off), for the persistence team, this essentially means how fast we can get persisted data back into dataflow on startup.

Let's write an end-to-end (i.e. including a dataflow) benchmark of our read performance to get a sense for how fast we can get data from disk into dataflow (spoiler: not very fast yet) and to use in tuning this path. I imagine this roughly as follows:

- Using our File storage (as opposed to Mem)
- A single persisted stream/collection
- Before the benchmark starts, write a bunch of data and seal it off (advance the timestamp). Persist can deal with any data type that implements persist::Codec (which includes Row), but we can avoid the dep on repr if we use `Vec<u8>` directly.
- This is interesting to vary in 2 particular dimensions: number of records and total size of data. So I think this ends up being 2, maybe 3, very related benchmarks. One that uses very small records (maybe a serialized u64 or u32 key with no value) to explore our per-record overhead. Another that uses records that are so big that we shouldn't see much per-record overhead (4KB/32KB?). Possibly a third that uses a "representative" key size (no idea, 100B?)
- In the benchmark, we build a dataflow (it's important to include dataflow construction in the benchmark timing so we can't cheat by reading stuff when creating the operator) that does something very simple with the output of the PersistedSource operator. Ideally, this would be something that, e.g. if we're mmap'ing all the various files involved, we could actually do without reading all the large values off disk. Maybe map over each, decoding the first bytes of each key as a u64 (which is sorta similar to how this would work with Row), and computing if it's even or odd?
- The PersistedSource also tails any new writes, so it won't drop its capabilities after it finishes reading the previous data. You'll probably want to use a probe to figure out when it's caught up and end the benchmark based on that.

I'm sure I've forgotten some things, but hopefully this is enough of a sketch to get started. It will also help you familiarize yourself with the external interface of src/persist, which should help with the source integration stuff.

**Comment by aljoscha** [comment]
I'm guessing we want the benchmark to only use vanilla timely dataflow (or differential dataflow at the most), and no materialize dataflow/row business?

**Comment by danhhz** [comment]
Yup! Almost certainly vanilla differential dataflow. I've been trying to keep the persist crate API mostly independent of mz (though the functionality & requirements driven purely by what mz needs) as a forcing function to help keep a strong API boundary. Also for very low deps and fast compiles :)
